feat(hyper): switch to std::io, std::net, and std::path.

seanmonstar opened the pull request:
All instances of `old_io` and `old_path` were switched to use the new
shiny `std::io`, `std::net`, and `std::path` modules. This means that
`Request` and `Response` implement `Read` and `Write` now.

Because of the changes to `TcpListener`, this also takes the opportunity
to correct the method usage of `Server`. As with other
languages/frameworks, the server is first created with a handler, and
then a host/port is passed to a `listen` method. This reverses what
`Server` used to do.

Closes #347

BREAKING CHANGE: Check the docs. Everything was touched.

reem commented:
Needs a rebase already.

reem commented on Cargo.toml:
```diff
@@ -21,3 +21,6 @@ time = "*"
 url = "*"
+
+[profile.bench]
+opt-level = 0
```
??

seanmonstar replied:
Oh dangit. I ahead set that when I was fixing the bench marks, to compile faster.

seanmonstar commented:
@reem rebased.

Interesting to note: speed improvements with the `client_mock_tcp` bench is over 500%.

```
BEFORE
test bench_mock_hyper ... bench:    280059 ns/iter (+/- 62609)

AFTER
test bench_mock_hyper ... bench:     50152 ns/iter (+/- 5771)
```

reem commented on src/http.rs:
```diff
@@ -65,13 +65,13 @@ impl<R: Reader> HttpReader<R> {
                 debug!("Sized read, remaining={:?}", remaining);
                 if *remaining == 0 {
-                    Err(old_io::standard_error(old_io::EndOfFile))
+                    Ok(0)
```
I think `Ok(0)` is still bad behavior, unless the recent RFC changed that.

seanmonstar replied:
My understanding is that Ok(0) is EOF. There no longer is an EOF error kind...

reem replied:
Asked about this and you're right.

reem commented on src/http.rs:
```diff
@@ -266,29 +287,24 @@ impl<W: Writer> Writer for HttpWriter<W> {
+                if msg.len() != 0 {
+                    error!("Cannot include a body with this kind of message");
                 }
+                Ok(0)
```
Same `Ok(0)` note here.

reem commented on src/http.rs:
```diff
@@ -345,24 +361,36 @@ pub fn is_token(b: u8) -> bool {
-fn read_token_until_space<R: Reader>(stream: &mut R, buf: &mut [u8]) -> HttpResult<bool> {
-    use std::old_io::BufWriter;
-    let mut bufwrt = BufWriter::new(buf);
+fn read_token_until_space<R: Read>(stream: &mut R, buf: &mut [u8]) -> HttpResult<bool> {
```
can this be renamed, since it is specialized to methods?

reem commented on src/http.rs:
````diff
@@ -507,14 +549,24 @@ pub type RawHeaderLine = (String, Vec<u8>);
 /// > ```
-pub fn read_header<R: Reader>(stream: &mut R) -> HttpResult<Option<RawHeaderLine>> {
+pub fn read_header<R: Read>(stream: &mut R) -> HttpResult<Option<RawHeaderLine>> {
+    macro_rules! byte (
````
This whole macro business is honestly horrible, why is this necessary? Can't we add an extension method?

seanmonstar replied:
Yea, it's gross. The change from an EOF error kind to Ok(0) makes this harder. I tried a ReadByte extension trait, but could propagate the correct error upon EOF. So, the macros allow inserting the appropriate HttpError when EOF is found early. 

I also have been working on a branch to use httparse, which removes much of this module.

reem replied:
Just found httpparse, seems interesting but we'll need a push parser for async integration later.

seanmonstar replied:
httparse is a push parser. My goal is to be like Pico, but using Iterator,
to keep things safe.

On Sun, Mar 1, 2015, 6:51 PM Jonathan Reem <notifications@github.com> wrote:

> In src/http.rs
> <https://github.com/hyperium/hyper/pull/354#discussion_r25576303>:
>
> > @@ -507,14 +549,24 @@ pub type RawHeaderLine = (String, Vec<u8>);
> >  /// >                ; obsolete line folding
> >  /// >                ; see Section 3.2.4
> >  /// > ```
> > -pub fn read_header<R: Reader>(stream: &mut R) -> HttpResult<Option<RawHeaderLine>> {
> > +pub fn read_header<R: Read>(stream: &mut R) -> HttpResult<Option<RawHeaderLine>> {
> > +    macro_rules! byte (
>
> Just found httpparse, seems interesting but we'll need a push parser for
> async integration later.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/hyperium/hyper/pull/354/files#r25576303>.
>

reem replied:
Oh, so the idea is I call `parse` many times?

seanmonstar replied:
Yes, but just like Pico, it's stateless. It will update the `Request` object with all the info it could find, but it won't keep track of it's state (as Pico's developer points out, keeping state makes it slower). So you can read some bytes, call `parse()`, and if it came back as `Status::Impartial` (or `-2` in Pico), you can still inspect some properties. Perhaps you check the route and can determine it's a 404 early before having to parse any more. You should, however, read more bytes into the same buffer, and call `parse()` again. The parser will start from the beginning.

It's faster this way because: keeping state (and checking it every single byte) is slower, and in **most** cases, reading from a socket should read the entire head, so you shouldn't often get `Partial` parses.

seanmonstar commented:
If there's no real problems, I'd like to merge so we can unbreak people on crates.io...

akoskaaa commented:
Hey, I am not sure if this is a blocker, but with the current build the example client does not build for me. Here's the exception:
```
src/main.rs:19:20: 19:36 error: type `hyper::client::response::Response` does not implement any method in scope named `read_to_string`
src/main.rs:19     let body = res.read_to_string().unwrap();
                                  ^~~~~~~~~~~~~~~~
src/main.rs:19:36: 19:36 help: methods from traits can only be called if the trait is in scope; the following trait is implemented but not in scope, perhaps add a `use` for it:
src/main.rs:19:36: 19:36 help: candidate #1: use `std::io::Read`
error: aborting due to previous error
```

maximih commented:
@akoskaaa read_to_string() was changed in the new io mdoule. Looks like the example needs to be updated. Here is how I modified my code to make it compile:
        
        use std::io::Read;
        ...
        let mut content: String = "".to_string();
        try!(response.read_to_string(&mut content));

seanmonstar commented:
Alright, I updated the REAME to work, and renamed that odd function. Once travis hits green, I'll merge. (sorry those using git rev's, I probably broke you because of my force-push)
